Replace os_sockQueryInterfaces by os_getifaddrs

k0ekk0ek opened the pull request:
This pull request splits requesting the available network interfaces from the socket abstraction. This makes it easier to implement support for alternative BSD socket compatible implementations that offer the same API (e.g. lwIP), but require the user to query the available network interfaces in a different way (which is more-or-less tied to the operating system and not necessarily the IP stack).

The API is loosely inspired by the getifaddrs interface found on modern Unices.

This pull requests also adds a platform independent implementation of asprintf and a function named memdup to easily copy n bytes of memory. Apart from that it also removes some duplicate function implementations.

kydos commented:
@k0ekk0ek, please fix the failing ip-validation checks and continuous integration.

eboasson commented on src/os/src/windows/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,260 @@
+    assert(addrtable != NULL);
+    assert(addr != NULL);
+
+    if ((ifa = os_calloc_s(1, sizeof(*ifa))) == NULL) {
```
First and only use of os_calloc in the code base — calloc is standard so I can live with it, but I'm not sure having it in is worth it if it is used so rarely ...

k0ekk0ek replied:
It's been available since the initial code submission. I guess nobody found it yet? For now I'd like to leave this in if you don't mind(?)

k0ekk0ek commented on src/os/src/windows/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,260 @@
+    DWORD ret;
+    size_t i;
+
+    static const size_t max = 2;
```
Once to get the required buffer size and once to actually copy in the table. (3 is a remnant from the old code, I'll change that to be 2 as well).

k0ekk0ek commented on src/os/src/windows/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,260 @@
+    ULONG ret;
+    size_t i;
+
+    static const size_t max = 3;
```
Will change it to 2.

k0ekk0ek commented on src/core/ddsi/include/ddsi/ddsi_ipaddr.h:
```diff
@@ -17,7 +17,7 @@
 enum ddsi_locator_from_string_result ddsi_ipaddr_from_string (ddsi_tran_factory_t tran, nn_locator_t *loc, const char *str, int32_t kind);
 char *ddsi_ipaddr_to_string (ddsi_tran_factory_t tran, char *dst, size_t sizeof_dst, const nn_locator_t *loc, int with_port);
-void ddsi_ipaddr_to_loc (nn_locator_t *dst, const os_sockaddr_storage *src, int32_t kind);
+void ddsi_ipaddr_to_loc (nn_locator_t *dst, const os_sockaddr *src, int32_t kind);
```
Replaced os_sockaddr_storage in these calls and removed some unused + unwanted functions while I was at it.

k0ekk0ek commented on src/core/ddsi/src/ddsi_raweth.c:
```diff
@@ -347,42 +347,23 @@ static void ddsi_raweth_deinit(void)
-    cnt++;
+
+  if ((err = os_getifaddrs(interfs, &filt)) == 0) {
+    for (ifa = *interfs; ifa != NULL; ifa = ifa->next, cnt++) {
```
Agreed. I wanted to keep the number of changes as small as possible, but it's resolved now.

k0ekk0ek commented on src/core/ddsi/src/ddsi_tran.c:
```diff
@@ -310,10 +310,10 @@ char *ddsi_locator_to_string_no_port (char *dst, size_t sizeof_dst, const nn_loc
-int ddsi_enumerate_interfaces (ddsi_tran_factory_t factory, int max, struct os_ifAttributes_s *interfs)
+int ddsi_enumerate_interfaces (ddsi_tran_factory_t factory, os_ifaddrs_t **interfs)
 {
   /* FIXME: HACK */
```
Processed your advice.

k0ekk0ek commented on src/os/src/posix/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,134 @@
+             || (addr->sa_family == AF_PACKET && ifltr->af_packet)
+#endif /* __linux */
+             || (addr->sa_family == AF_INET6 && ifltr->af_inet6 &&
+                 !IN6_IS_ADDR_UNSPECIFIED(
```
I also opted for 2.

k0ekk0ek commented on src/os/src/posix/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,112 @@
+                err = errno;
+            }
+        } else {
+            assert(sys_ifa->ifa_netmask != NULL);
```
Changed it.

k0ekk0ek commented on src/os/src/posix/os_platform_ifaddrs.c:
```diff
@@ -0,0 +1,112 @@
+        ifa = ifa_root = NULL;
+
+        for (sys_ifa = sys_ifa_root;
+             sys_ifa != NULL && sys_ifa->ifa_addr != NULL && err == 0;
```
Should be good now.

k0ekk0ek commented:
Removed some more unused + unwanted functions, consolidated some others and added some extra test cases. Tests are running fine on my Windows 10 machine, but they still appear to cause trouble in the AppVeyor builds. I believe I even noticed a segmentation fault? I'll wait for AppVeyor to complete and then investigate again. I couldn't trigger the behaviour even after disabling IPv6 on my Windows 10 machine...

k0ekk0ek commented:
@eboasson, AppVeyor build is failing due to timeout. Probably due to Criterion? Do we consider this good enough for now? I'm planning on removing Criterion next. Do we want to wait for that?

eboasson commented:
@k0ekk0ek I'm ok with it: the symptoms really match the known issue with Criterion, and in this particular case there is furthermore no clear reason why this one test that timed would have been affected by these changes.
